sdk/dotnet/Pulumi/Serialization/Serializer.cs

```diff
@@ -134,25 +134,56 @@ public Serializer(bool excessiveDebugOutput)
                 {
```

Comment:
I was wondering what is the concern here. Would anything notice the difference if the serialized form has a `v` or an empty output? Is this just to optimize the payload size a bit?

If we could formalize equivalence "up to autoconvert" for `Output(v, known=True, deps=[], secret=False) ~ v` we could be tolerant when deserializing, and not care in which shape the things are serialized, would that be simpler? Or I am missing something that truly cares which form it is?

Reply:
> Is this just to optimize the payload size a bit?

Right. It's not really a problem, it's just a little unfortunate that every single value ends up as an output value. In any case, I went ahead and added a commit that avoids emitting an output value when it's not needed.

More problematic: I've found that in the .NET SDK `Input<T>` (and `InputList<T>`, `InputMap<V>`, and `InputUnion<T0, T1>`) eagerly stores the value of `T` as an `Output<T>`:

https://github.com/pulumi/pulumi/blob/0a38bc295cfc5ae3c325dcece169da5a5e80fffc/sdk/dotnet/Pulumi/Core/Input.cs#L22-L31

So a test like this fails:

```csharp
[Fact]
public static Task TestSecretInInputList()
    => RunInNormal(async () =>
    {
        var input = new InputList<string>{ Output.CreateSecret("hello") };
        var expected = ImmutableArray<object>.Empty.Add(
            ImmutableDictionary<string, object>.Empty
                .Add(Constants.SpecialSigKey, Constants.SpecialOutputValueSig)
                .Add("value", "hello")
                .Add("secret", true));

        var s = new Serializer(excessiveDebugOutput: false);
        var actual = await s.SerializeAsync("", input, keepResources: true, keepOutputValues: true);

        Assert.Equal(expected, actual!);
    });
```

I'm expecting the serialized value to look like (the value inside the list is an output value):

```json
[
    {
        "4dabf18193072939515e22adb298388d": "d0e6a833031e9bbcd3f4e8bde6ca49a4",
        "value": "hello",
        "secret": true
    }
]
```

But actually getting (the top-level value is an output value):

```json
{
    "4dabf18193072939515e22adb298388d": "d0e6a833031e9bbcd3f4e8bde6ca49a4",
    "value": ["hello"],
    "secret": true
}
```

Which basically defeats the purpose of why we added output values to begin with.

It turns out that regular .NET programs (that aren't even using MLCs) end up having undesirable previews because of this behavior with `Input<T>`. For example:

```csharp
using Pulumi;
using Pulumi.Random;
using Pulumi.Aws.S3;

class MyStack : Stack
{
    public MyStack()
    {
        var randString = new RandomString("str", new RandomStringArgs { Length = 10 });
        var bucket = new Bucket("bucket", new BucketArgs {
            Tags = {
                { "hello", "world" },
                { "hi", "there" },
                { "foo", randString.Id },
            }
        });
    }
}
```

Has the following preview diff:

```
        tags        : output<string>
```

When we want it to be like this:

```
        tags        : {
            hello     : "world"
            hi        : "there"
            foo       : output<string>
        }

```

My suggestion is we merge this PR as-is (modulo addressing any review feedback) and open an issue to track this issue with `Input<T>` eagerly converting values to `Output<T>`. And fix that separately. I'm thinking the fix will involve just avoiding/deferring the eager `Output<T>` conversion as much as possible, and only do it when needed.